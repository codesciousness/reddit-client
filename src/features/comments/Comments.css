.Comments {
    margin: 4rem auto auto auto;
}

.Comments__error {
    margin: 2rem 0;
}

.Comments__link, .Comments__link:hover, .Comments__link:visited {
    color: steelblue;
    text-decoration: none;
}

.Comments__icon {
    font-size: 1.5rem;
}

.Comments li {
    list-style: none;
}