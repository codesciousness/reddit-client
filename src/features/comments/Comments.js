import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import './Comments.css';
import Comment from '../../components/comment/Comment';
import Loader from '../../components/loader/Loader';
import { selectCurrentPost } from '../posts/postsSlice';
import { selectComments, selectLoadingComments, selectLoadCommentsError, loadComments } from './commentsSlice';

const Comments = () => {
    const comments = useSelector(selectComments);
    const loadingComments = useSelector(selectLoadingComments);
    const loadCommentsError = useSelector(selectLoadCommentsError);
    const currentPost = useSelector(selectCurrentPost);
    const dispatch = useDispatch();

    useEffect(() => {
        if (currentPost) {
            dispatch(loadComments(currentPost));
        }
    }, [currentPost, dispatch]);

    if (loadingComments) {
        return (
            <section className="Comments">
                <h3 className="Comments__title">Comments</h3>
                <Loader />
            </section>
        );
    }
    if (loadCommentsError) {
        return (
            <section className="Comments">
                <h3 className="Comments__title">Comments</h3>
                <p className="Comments__error">Error: Request failed! Please try again. ¯\_(ツ)_/¯</p>
                <p className="Comments__error">Or Return to <Link to='/' className="Comments__link"><i className="fas fa-home Comments__icon"></i> Home</Link></p>
            </section>
        );
    }
    return (
        <section className="Comments">
            <h3 className="Comments__title">Comments</h3>
            {comments.map(comment => <Comment comment={comment} />)}
        </section>
    );
}

export default Comments;