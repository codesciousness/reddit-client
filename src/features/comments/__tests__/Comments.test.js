import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { BrowserRouter as Router } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import { create } from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import Comments from '../Comments';

const mockStore = configureStore([thunk]);

const comments = [
    {
        id: 'abcd0e',
        author: 'ScienceReader',
        created: '1 hour ago',
        body: 'Very informative article!',
        ups: 10
    },
    {
        id: 'fghi1j',
        author: 'ScienceLearner',
        created: '15 minutes ago',
        body: 'Future research on this topic should definitely move more in this direction.',
        ups: 1
    }
];

describe('<Comments />', () => {
    let store;

    beforeEach(() => {
        store = mockStore({
            posts: {
                currentPost: 'jklm3n'
            },
            comments: {
                comments,
                loadingComments: false,
                loadCommentsError: false
            }
        });
    });

    test('Comments component is defined', () => {
        expect(<Comments />).toBeDefined();
    });
    test('matches the snapshot', () => {
        const component = create(
            <Provider store={store}>
                <Router>
                    <Comments />
                </Router>
            </Provider>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
    test('renders correctly', () => {
        render(
            <Provider store={store}>
                <Router>
                    <Comments />
                </Router>
            </Provider>
        );
        expect(screen.getByText('Comments')).toBeInTheDocument();
    });
    test('renders and passes each comment from state as a prop to a Comment component', () => {
        render(
            <Provider store={store}>
                <Router>
                    <Comments />
                </Router>
            </Provider>
        );
        expect(screen.getByText(/Very informative article/)).toBeInTheDocument();
        expect(screen.getByText(/ScienceLearner/)).toBeInTheDocument();
    });
    test('dispatches correct payload to async action creator when currentPost is truthy', () => {
        render(
            <Provider store={store}>
                <Router>
                    <Comments />
                </Router>
            </Provider>
        );
        const actions = store.getActions();
        const dispatchedPayload = actions[0].meta.arg;
        const expectedPayload = 'jklm3n';
        expect(dispatchedPayload).toBe(expectedPayload);
    });
    test('renders 2 CommentLoader components when loadingComments=true', () => {
        store = mockStore({
            posts: {
                currentPost: 'jklm3n'
            },
            comments: {
                comments,
                loadingComments: true,
                loadCommentsError: false
            }
        });
        const { container } = render(
            <Provider store={store}>
                <Router>
                    <Comments />
                </Router>
            </Provider>
        );
        expect(container.querySelectorAll('.CommentLoader')).toHaveLength(2);
    });
    test('renders error message when loadCommentsError=true', () => {
        store = mockStore({
            posts: {
                currentPost: 'jklm3n'
            },
            comments: {
                comments,
                loadingComments: false,
                loadCommentsError: true
            }
        });
        const { container } = render(
            <Provider store={store}>
                <Router>
                    <Comments />
                </Router>
            </Provider>
        );
        expect(container.querySelector('.Comments__error')).toBeInTheDocument();
        expect(container.querySelector('.Comments__link')).toBeInTheDocument();
        expect(screen.getByText(/Error: Request failed! Please try again./)).toBeInTheDocument();
    });
});