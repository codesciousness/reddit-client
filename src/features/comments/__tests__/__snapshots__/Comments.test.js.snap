// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Comments /> matches the snapshot 1`] = `
<section
  className="Comments"
>
  <h3
    className="Comments__title"
  >
    Comments
  </h3>
  <ul>
    <li>
      <section
        className="Comment"
        id="abcd0e"
      >
        <p
          className="Comment__author"
        >
          ScienceReader
           | 
          <span
            className="PostTime"
          >
            NaN years ago
          </span>
        </p>
        <p
          className="Comment__body"
        >
          Very informative article!
        </p>
        <p
          className="Comment__ups"
        >
          <i
            className="fas fa-arrow-up"
          />
           
          10
           Ups
        </p>
      </section>
    </li>
    <li>
      <section
        className="Comment"
        id="fghi1j"
      >
        <p
          className="Comment__author"
        >
          ScienceLearner
           | 
          <span
            className="PostTime"
          >
            NaN years ago
          </span>
        </p>
        <p
          className="Comment__body"
        >
          Future research on this topic should definitely move more in this direction.
        </p>
        <p
          className="Comment__ups"
        >
          <i
            className="fas fa-arrow-up"
          />
           
          1
           Ups
        </p>
      </section>
    </li>
  </ul>
</section>
`;
